nodelifecycle controller: reconcile node OS/arch labels

yujuhong opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190130-node-os-arch-labels.md

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #72929

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

yujuhong committed:
nodelifecycle controller: reconcile node OS/arch labels

yujuhong commented:
/assign @k82cn 
/cc @liggitt @thockin

yujuhong commented:
/cc @dchen1107 @derekwaynecarr

tylermann commented on pkg/controller/controller_utils.go:
```diff
@@ -1045,3 +1051,48 @@ func ComputeHash(template *v1.PodTemplateSpec, collisionCount *int32) string {
+		}
+		newData, err := json.Marshal(newNode)
+		if err != nil {
+			return err
```
It might be worth using `fmt.Errorf` to add context to these errors before returning. Otherwise it could be hard to distinguish which error case was hit in this function since there are multiple of them. Eg:
```
return fmt.Errorf("failed to marshal new node %#v: %v", newNode, err)
```

yujuhong replied:
Done

tylermann commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -131,6 +131,43 @@ const (
+//   - If both keys exist, but their values don't match. Use the value from the
+//   primaryKey as the source of truth to reconcile.
+//   - If ensureSecondaryExists is true, and the secondaryKey does not
+//   exist, secondardayKey will be added with the value of the primaryKey.
```
minor typo here `secondardayKey` -> `secondaryKey`

yujuhong replied:
Fixed

tylermann commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -1192,6 +1240,53 @@ func (nc *Controller) ComputeZoneState(nodeReadyConditions []*v1.NodeCondition)
 	}
 }
 
+// reconcileNdoeLabels reconciles node labels.
```
minor typo here `reconcileNdoeLabels` -> `reconcileNodeLables`

yujuhong replied:
Done.

tylermann commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -1192,6 +1240,53 @@ func (nc *Controller) ComputeZoneState(nodeReadyConditions []*v1.NodeCondition)
+	}
+
+	if node.Labels == nil {
+		// Nothing to reconcile.
```
thanks for these inline comments, definitely improves readability here.

tylermann reviewed (commented):
Was just reading over to learn some about the project and left some comments along the way. In general this looks good to me, nice work.

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -131,6 +131,43 @@ const (
 
+const (
+	// These labels MUST match what defined in pkg/kubelet/apis/well_known_labels.go
+	labelOSBeta   = "beta.kubernetes.io/os"
```
prefer to import "pkg/kubelet/apis" here.

yujuhong replied:
Done

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -446,10 +485,19 @@ func (nc *Controller) doNoScheduleTaintingPassWorker() {
+
+		// Requeue the node if needed.
+		if needsToRequeue {
+			nc.nodeUpdateQueue.Add(nodeName)
```
If add it back to queue, it's better to use working queue with ratelimit.

yujuhong replied:
Done.

k82cn replied:
For ratelimit working queue, we need to use `AddRateLimited` and call `Forget(req)`  if succeed.

yujuhong replied:
You're right. Sorry for missing that. Fixed in the new commit

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -446,10 +485,19 @@ func (nc *Controller) doNoScheduleTaintingPassWorker() {
 		nodeName := obj.(string)
-
+		needsToRequeue := false
 		if err := nc.doNoScheduleTaintingPass(nodeName); err != nil {
```
we need to check `TaintNodeByCondition` flag here, it's beta feature right now.

yujuhong replied:
Added in the new commit

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -446,10 +485,19 @@ func (nc *Controller) doNoScheduleTaintingPassWorker() {
 			klog.Errorf("Failed to taint NoSchedule on node <%s>, requeue it: %v", nodeName, err)
+			needsToRequeue = true
+		}
+		if err := nc.reconcileNodeLables(nodeName); err != nil {
```
It's better to add a TODO here, e.g. we'll remove this after 1.18.

yujuhong replied:
We still need this in 1.18 (see the KEP). There are also other labels under discussion that are may rely this mechanism. I'll add a TODO to re-evaluate in 1.19

k82cn replied:
ok to me :)

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -131,6 +131,43 @@ const (
+var labelReconcileInfo = []struct {
+	primaryKey            string
+	secondaryKey          string
+	ensureSecondaryExists bool
```
if `ensureSecondaryExists` is only one value in this PR, prefer to remove it for now; and add it back when there're other cases.

yujuhong replied:
We are going to use this mechanism, and I think leaving this actually makes the code more readable when we switch in the future.

If you're okay, I'd rather leave this as it is.

k82cn replied:
that's ok

k82cn commented on pkg/controller/controller_utils.go:
```diff
@@ -1045,3 +1051,48 @@ func ComputeHash(template *v1.PodTemplateSpec, collisionCount *int32) string {
 	return rand.SafeEncodeString(fmt.Sprint(podTemplateSpecHasher.Sum32()))
 }
+
+func AddOrUpdateLabelsOnNode(kubeClient clientset.Interface, nodeName string, labelsToUpdate map[string]string) error {
```
Is it possible to combine "the update call to API server" with `TaintByCondition` ? If we want to it later because of 1.14 schedule, please help to add a TODO or open an issue for that :)

yujuhong replied:
The `TaintByCondition` mechanism is more complex (i.e., swapping taints) and is still gated. I think merging them may simply cause unnecessary bugs. Label reconciling happens rarely (e.g., node upgrade/downgrade with mismatching OS/arch) , and having a separate apiserver update seems okay to me.

k82cn replied:
I'm ok to keep it as is.

we can have seperate PR for that; as prefer to have such logic : get node info, reconcile it (e,g. labels, conditions), update it back to api server. If we want to add other reconcile logic, we only need to enhance the seconds phase.

yujuhong replied:
Sounds good.

yujuhong commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -440,18 +435,16 @@ func (nc *Controller) Run(stopCh <-chan struct{}) {
 		go nc.taintManager.Run(stopCh)
 	}
 
-	if nc.taintNodeByCondition {
```
@k82cn I moved the condition to inside the worker (line 481) since we do want to reconcile node labels even if taintNodeByCondition is false.

yujuhong reviewed (commented):
Thanks for reviewing.

Addressed comments (in the new commit). PTAL again!

yujuhong commented:
/test pull-kubernetes-integration

yujuhong commented:
/test pull-kubernetes-e2e-gce

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -390,11 +390,11 @@ func NewNodeLifecycleController(
 		}),
 		UpdateFunc: nodeutil.CreateUpdateNodeHandler(func(_, newNode *v1.Node) error {
-			nc.nodeUpdateQueue.Add(newNode.Name)
+			nc.nodeUpdateQueue.AddRateLimited(newNode.Name)
```
IMO, `Add` is for normal case, `AddRateLimited` for retry :)

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -493,9 +493,9 @@ func (nc *Controller) doNodeProcessingPassWorker() {
 
```
add `defer nc.nodeUpdateQueue.Done(nodeName)` here.

k82cn commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -493,9 +493,9 @@ func (nc *Controller) doNodeProcessingPassWorker() {
 		// Requeue the node if needed.
 		if needsToRequeue {
-			nc.nodeUpdateQueue.Add(nodeName)
+			nc.nodeUpdateQueue.AddRateLimited(nodeName)
```
add `continue` after `AddRateLimited`

yujuhong commented:
@k82cn I fixed the rate-limiting queue usage, and the tests are passing finally.
PTAL again. Thanks for reviewing!

@derekwaynecarr do you want to take a look and/or approve?

derekwaynecarr commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -1191,6 +1238,53 @@ func (nc *Controller) ComputeZoneState(nodeReadyConditions []*v1.NodeCondition)
 }
 
+// reconcileNodeLabels reconciles node labels.
+func (nc *Controller) reconcileNodeLables(nodeName string) error {
```
nit: typo

yujuhong replied:
What a wonderfully consistent typo that I made everywhere.

derekwaynecarr commented:
@yujuhong the code looks good, just the one minor typo that we can address later.

/approve
/lgtm

yujuhong reviewed (commented):
Fixed the typos and squashed the commits.
Thanks for reviewing!

yujuhong commented:
Rebased (no conflict)

yujuhong commented:
Re-applying lgtm based on https://github.com/kubernetes/kubernetes/pull/74442#issuecomment-470314246

yujuhong commented:
/test pull-kubernetes-e2e-gce-100-performance

yujuhong commented:
Tests using large clusters are failing most likely because of the introduction of rate limiter. I bumped the values to see if it helps.

@k82cn @derekwaynecarr 

I added the rate limiter based on the review feedback, but in reality, label reconciling happens way less frequently than taintNodeByCondition (which is what caused the failure). 
I'd like to revert the changes (i.e., remove the rate limiters) if we are not sure about what the right limits are.

@k82cn reverting that change would also mean removing the "requeue" operation I added in the PR. How does "taintNodeByCondition" handle failures?

yujuhong commented:
> Tests using large clusters are failing most likely because of the introduction of rate limiter. I bumped the values to see if it helps.

Bumping the limits seems to have solved the issue for the 100-node cluters. 

I'm going to revert the introduction of rate-limiting queue in this PR since we need more measurements for setting proper limits (especially for taintNodeByCondition).

k82cn commented:
/lgtm , thanks :)

yujuhong commented:
/retest

yujuhong commented:
/test pull-kubernetes-e2e-gce

All tests passed but the job timed out during cluster teardown? Trying again to see if it's just flaky

yujuhong commented:
/test pull-kubernetes-e2e-gce

yujuhong commented:
> /lgtm , thanks :)

I guess bot could not recognize this. Applied the lgtm manually. Thanks everyone for reviewing the PR.
